apps/emqx_connector/src/emqx_connector_template.erl

```diff
@@ -0,0 +1,346 @@
+-type deep() :: {'$tpl', deeptpl()}.
+
+-type deeptpl() ::
+    t()
+    | #{deeptpl() => deeptpl()}
+    | {list, [deeptpl()]}
+    | {tuple, [deeptpl()]}
+    | scalar()
+    | function()
+    | pid()
+    | port()
+    | reference().
+
+-type placeholder() :: {var, varname(), accessor()}.
+-type accessor() :: [binary()].
+-type varname() :: string().
+
+-type scalar() :: atom() | unicode:chardata() | number().
+-type binding() :: scalar() | list(scalar()) | bindings().
+-type bindings() :: #{atom() | binary() => binding()}.
+
+-type var_trans() ::
+    fun((Value :: term()) -> unicode:chardata())
+    | fun((varname(), Value :: term()) -> unicode:chardata()).
+
+-type parse_opts() :: #{
+    strip_double_quote => boolean()
+}.
+
+-type render_opts() :: #{
+    var_trans => var_trans()
+}.
+
+-define(RE_PLACEHOLDER, "\\$(\\$?)\\{[.]?([a-zA-Z0-9._]*)\\}").
+
+%% @doc Parse a unicode string into a template.
+%% String might contain zero or more of placeholders in the form of `${var}`,
+%% where `var` is a _location_ (possibly deeply nested) of some value in the
+%% bindings map.
+%% String might contain special escaped form `$${...}` which interpreted as a
+%% literal `${...}`.
+-spec parse(String :: unicode:chardata()) ->
+    t().
+parse(String) ->
+    parse(String, #{}).
+
+-spec parse(String :: unicode:chardata(), parse_opts()) ->
+    t().
+parse(String, Opts) ->
+    RE =
+        case Opts of
+            #{strip_double_quote := true} ->
+                <<"((?|" ?RE_PLACEHOLDER "|\"" ?RE_PLACEHOLDER "\"))">>;
+            #{} ->
+                <<"(" ?RE_PLACEHOLDER ")">>
+        end,
+    Splits = re:split(String, RE, [{return, binary}, group, trim, unicode]),
+    Components = lists:flatmap(fun parse_split/1, Splits),
+    Components.
+
+parse_split([Part, _PH, <<>>, Var]) ->
+    % Regular placeholder
+    prepend(Part, [{var, unicode:characters_to_list(Var), parse_accessor(Var)}]);
+parse_split([Part, _PH = <<B1, $$, Rest/binary>>, <<"$">>, _]) ->
+    % Escaped literal, take all but the second byte, which is always `$`.
+    % Important to make a whole token starting with `$` so the `unparse/11`
+    % function can distinguish escaped literals.
+    prepend(Part, [<<B1, Rest/binary>>]);
+parse_split([Tail]) ->
+    [Tail].
+
+prepend(<<>>, To) ->
+    To;
+prepend(Head, To) ->
+    [Head | To].
+
+parse_accessor(Var) ->
+    case string:split(Var, <<".">>, all) of
+        [<<>>] ->
+            ?PH_VAR_THIS;
+        Name ->
+            % TODO: lowercase?
+            Name
+    end.
+
+-spec validate([varname()], t()) ->
+    ok | {error, [_Error :: {varname(), disallowed}]}.
+validate(Allowed, Template) ->
+    {_, Errors} = render(Template, #{}),
+    {Used, _} = lists:unzip(Errors),
+    case lists:usort(Used) -- Allowed of
+        [] ->
+            ok;
+        Disallowed ->
+            {error, [{Var, disallowed} || Var <- Disallowed]}
+    end.
+
+-spec trivial(t()) ->
```

Comment:
add a `@doc` ?

Reply:
Added in [0eaf351](https://github.com/emqx/emqx/pull/10442/commits/0eaf351055c9adaf796e8ab849f311e469d1b283) (more missing `@doc`s in [d1ab95d](https://github.com/emqx/emqx/pull/10442/commits/d1ab95da3576db9095ccdb27e8ab457708f83534)).